API back REPL

libby opened the pull request:
Adding a REPL that is back by the Juno API (http calls) and outputs the JSON response. To run `stack exec -- junorepl`
Supports the normal commands: 
`$> ObserveAccounts`
`$> CreateAccount Acct1`
`$> AdjustAccount Acct1 100.0`
`$> Poll {requestId}`
`$> transfer(Acct1->Acct2, 1%1)`
and batching, submitting a batch of size X of the supplied command.
`$> batch test: 500 transfer(Acct1->Acct2, 1 % 1)`

libbylite committed:
Adding a REPL that is backed by the API, REPL command make http requests behind the scenes.

For now the API REPL doesn't support the many test, and outputs the full JSON response from the server.

libbylite committed:
Adding jmeter.log file to .gitignore

libbylite committed:
Adding support for sumbitting batches from the API backed REPL.

To submit a batch from the API backed REPL:
batch test: 500 AdjustAccount Acct1 2.0
batch test: 500 AdjustAccount Acct1 2.0

libbylite committed:
Removing test print statement.

libbylite committed:
Adding exception handling to API backed REPL.

Wrapping the API REPL in catch so that it doesn't break on invalid input.

sirlensalot commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+    "" -> runREPL
+    _ -> catch
+         (processInput cmd)
+         (\e -> do
```
use ScopedTypeVariables, ie `catch (...) (\(e :: SomeException) -> ...`

slpopejoy commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+              putStrLn $ "Invalid Command: " ++ err
+         ) >> runREPL
+  where
+    processInput input = do
```
move BSC.pack up into calling context, ie `processInput (BSC.pack cmd)`

slpopejoy commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+      -- batch test: 500 AdjustAccount Acct1 2.0
+      if take 11 input == batchToken
+      then do
+        let batchCmd = T.strip . T.pack $ drop 11 input
```
write a parser

slpopejoy commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+        showResult res
+      else if take 4 input == "Poll"
+      then do
+        let cmds = (tail . words) input
```
ick

slpopejoy commented on src/Apps/Juno/JsonTypes.hs:
```diff
@@ -292,3 +293,15 @@ mJsonBsToCommand bs = JSON.decode bs >>= \v ->
 
     mtransact (TransactBody code _) =
         BSC.pack $ T.unpack code
+
```
1) move to Repl.hs
2) write a parser

sirlensalot commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+      cmd' <- return $ BSC.pack input
+      -- batch test: 500 transfer(Acct1->Acct2, 1 % 1)
+      -- batch test: 500 AdjustAccount Acct1 2.0
+      if take 11 input == batchToken
```
1) break into function 2) use a parser

slpopejoy commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+        let jsonBytes = cmdBatch2JSON $ JsonT.CommandBatch txBatch
+        res <- submitCmdBatch jsonBytes
+        showResult res
+      else if take 4 input == "Poll"
```
parser

sirlensalot commented on src/Apps/Juno/Repl.hs:
```diff
@@ -0,0 +1,110 @@
+        case readHopper cmd' of
+          Left err -> putStrLn input >> putStrLn err >> runREPL
+          Right _ -> do
+             let cmdJsonBytes = T.pack . BLC.unpack $ JsonT.commandToJSONBytes cmd'
```
decodeUtf8

sirlensalot commented on src/Apps/Juno/JsonTypes.hs:
```diff
@@ -292,3 +293,15 @@ mJsonBsToCommand bs = JSON.decode bs >>= \v ->
+      _ -> JSON.encode $ TransactBody (E.decodeUtf8 cmd) ""
+
+commandTextToJSONText :: T.Text -> T.Text
+commandTextToJSONText cmd = T.pack . BLC.unpack $ commandToJSONBytes $ BSC.pack . T.unpack $ cmd
```
decodeUtf8/encodeUtf8

sirlensalot commented on src/Apps/Juno/JsonTypes.hs:
```diff
@@ -292,3 +293,15 @@ mJsonBsToCommand bs = JSON.decode bs >>= \v ->
+commandToJSONBytes cmd =
+    case BSC.words cmd of
+      (cmd':acct:amount:[]) | cmd' == "AdjustAccount" ->
+        JSON.encode $  AccountAdjustPayload (E.decodeUtf8 acct) (read . T.unpack . E.decodeUtf8 $ amount)
```
BSC.unpack

sirlensalot commented on src/Apps/Juno/JsonTypes.hs:
```diff
@@ -292,3 +293,15 @@ mJsonBsToCommand bs = JSON.decode bs >>= \v ->
     mtransact (TransactBody code _) =
         BSC.pack $ T.unpack code
+
+commandToJSONBytes :: BSC.ByteString -> BLC.ByteString
```
Move into Repl

sirlensalot commented on src/Apps/Juno/JsonTypes.hs:
```diff
@@ -292,3 +293,15 @@ mJsonBsToCommand bs = JSON.decode bs >>= \v ->
+        JSON.encode $  AccountPayload (E.decodeUtf8 acct)
+      _ -> JSON.encode $ TransactBody (E.decodeUtf8 cmd) ""
+
+commandTextToJSONText :: T.Text -> T.Text
```
Move into Repl

libbylite committed:
Cleaning up API REPL code, making AdjustAccount take a Rational not a Double.

Break API REPL code into smaller functions, reuse hopperlite parser, add comments,
general cleanup. Change AdjustAccount to take a Rational not a double (update tests + jmeter).
TODO: Aeson json parser for Rational is blah `{numerator:,denominator:}` make a Json type class
parser to support `n%d`.

libbylite committed:
Adding an aeson json Typeclass instance for `Ratio Integer`.

The typeclass provided by aeson translates
`Ratio Integer -> { numerator: n, denominator: d }` now parsing
`Ratio Integer -> "n%d"` updated the tests as well.
